File: src/main/java/gvlfm78/plugin/OldCombatMechanics/module/ModuleSwordBlocking.java
package kernitus.plugin.OldCombatMechanics.module;

import kernitus.plugin.OldCombatMechanics.OCMMain;
import kernitus.plugin.OldCombatMechanics.utilities.Config;
import kernitus.plugin.OldCombatMechanics.utilities.ConfigUtils;
import kernitus.plugin.OldCombatMechanics.utilities.RunnableSeries;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Item;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.block.Action;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent.DamageModifier;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.player.*;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.PlayerInventory;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.*;

/**
 * Created by Rayzr522 on 7/4/16.
 */
public class ModuleSwordBlocking extends Module {

    private static final ItemStack SHIELD = new ItemStack(Material.SHIELD);
    private final Map<UUID, ItemStack> storedOffhandItems = new HashMap<>();
    private final Map<UUID, RunnableSeries> correspondingTasks = new HashMap<>();
    private int restoreDelay;
    private String blockingDamageReduction;
    private boolean blacklist;
    private List<Material> noBlockingItems = new ArrayList<>();

    public ModuleSwordBlocking(OCMMain plugin){
        super(plugin, "sword-blocking");
    }

    @Override
    public void reload(){
        restoreDelay = module().getInt("restoreDelay", 40);
        blockingDamageReduction = module().getString("blockingDamageReduction", "1")
                .replaceAll(" ", "");
        blacklist = module().getBoolean("blacklist");
        noBlockingItems = ConfigUtils.loadMaterialList(module(), "noBlockingItems");
    }


    @EventHandler(priority = EventPriority.HIGHEST)
    public void onRightClick(PlayerInteractEvent e){
        if(e.getItem() == null) return;

        Action action = e.getAction();

        if(action != Action.RIGHT_CLICK_AIR && action != Action.RIGHT_CLICK_BLOCK) return;

        if(action == Action.RIGHT_CLICK_BLOCK && Config.getInteractiveBlocks().contains(e.getClickedBlock().getType()))
            return;

        Player p = e.getPlayer();
        World world = p.getWorld();

        if(!isEnabled(world)) return;

        UUID id = p.getUniqueId();

        if(p.isBlocking()){
            tryCancelTask(id);
        } else {
            ItemStack item = e.getItem();

            if(!isHoldingSword(item.getType()) || hasShield(p)) return;

            PlayerInventory inv = p.getInventory();

            boolean isANoBlockingItem = noBlockingItems.contains(inv.getItemInOffHand().getType());

            if(blacklist && isANoBlockingItem || !blacklist && !isANoBlockingItem) return;

            storedOffhandItems.put(id, inv.getItemInOffHand());

            inv.setItemInOffHand(SHIELD);
        }

        scheduleRestore(p);
    }

    @EventHandler
    public void onHotBarChange(PlayerItemHeldEvent e){
        tryCancelTask(e.getPlayer().getUniqueId());
        restore(e.getPlayer());
    }

    @SuppressWarnings("deprecation")
    @EventHandler(priority = EventPriority.HIGHEST)
    public void onHit(EntityDamageByEntityEvent e){
        Entity ent = e.getEntity();

        if(!(ent instanceof Player)) return;

        Player p = (Player) ent;

        if(isBlocking(p.getUniqueId())){
            //If it's a player blocking
            //Instead of reducing damage to 33% apply config reduction

            double damageReduction = e.getDamage(); //Reducing by this would mean blocking all damage

            if(blockingDamageReduction.matches("\\d{1,3}%")){
                //Reduce damage by percentage
                int percentage = Integer.parseInt(blockingDamageReduction.replace("%", ""));
                damageReduction = damageReduction * percentage / 100;
            } else if(blockingDamageReduction.matches("\\d+")){
                //Reduce by specified amount of half-hearts
                damageReduction = Integer.parseInt(blockingDamageReduction);
            } else damageReduction = 0;

            if(damageReduction < 0) damageReduction = 0;

            //Only reduce damage if they were hit head on, i.e. the shield blocked some of the damage
            if(e.getDamage(DamageModifier.BLOCKING) >= 0) return;

            //Also make sure reducing the damage doesn't result in negative damage
            e.setDamage(DamageModifier.BLOCKING, 0);

            if(e.getFinalDamage() >= damageReduction)
                e.setDamage(DamageModifier.BLOCKING, damageReduction * -1);

            //Make maximum reduction possible be up to amount specified in config

            if(!isSettingEnabled("shieldFullBlock")){
                double minDamage = module().getDouble("minimumDamage");
                if(e.getFinalDamage() < minDamage)
                    e.setDamage(minDamage);
            }


            debug("Damage reduced by: " + e.getDamage(DamageModifier.BLOCKING), p);
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onWorldChange(PlayerChangedWorldEvent e){
        restore(e.getPlayer());
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPlayerLogout(PlayerQuitEvent e){
        restore(e.getPlayer());
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPlayerDeath(PlayerDeathEvent e){
        if(!isBlocking(e.getEntity().getUniqueId())) return;

        Player p = e.getEntity();
        UUID id = p.getUniqueId();

        e.getDrops().replaceAll(item -> {

            if(item.getType().equals(Material.SHIELD))
                item = storedOffhandItems.remove(id);

            return item;
        });

        // No need for any restore tasks, we've already done that
        tryCancelTask(id);
        // Handle keepInventory = true
        restore(p);
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onPlayerSwapHandItems(PlayerSwapHandItemsEvent e){
        Player p = e.getPlayer();
        if(isBlocking(p.getUniqueId()))
            e.setCancelled(true);

    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onInventoryClick(InventoryClickEvent e){

        if(e.getWhoClicked() instanceof Player){
            Player p = (Player) e.getWhoClicked();

            if(isBlocking(p.getUniqueId())){
                ItemStack cursor = e.getCursor();
                ItemStack current = e.getCurrentItem();
                if(cursor != null && cursor.getType() == Material.SHIELD ||
                        current != null && current.getType() == Material.SHIELD){
                    e.setCancelled(true);
                    restore(p);
                }
            }
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onItemDrop(PlayerDropItemEvent e){
        Item is = e.getItemDrop();

        Player p = e.getPlayer();

        if(isBlocking(p.getUniqueId()) && is.getItemStack().getType() == Material.SHIELD){
            e.setCancelled(true);
            restore(p);
        }
    }

    private void scheduleRestore(final Player player){
        BukkitRunnable removeItem = new BukkitRunnable() {
            @Override
            public void run(){
                restore(player);
            }
        };
        removeItem.runTaskLater(plugin, restoreDelay);

        BukkitRunnable checkBlocking = new BukkitRunnable() {
            @Override
            public void run(){
                if(!player.isBlocking()){
                    restore(player);
                    tryCancelTask(player.getUniqueId());
                }
            }
        };
        checkBlocking.runTaskTimer(plugin, 10L, 2L);

        correspondingTasks.put(player.getUniqueId(), new RunnableSeries(removeItem, checkBlocking));
    }

    private void restore(Player p){
        UUID id = p.getUniqueId();

        if(!isBlocking(id)) return;

        if(p.isBlocking()) //They are still blocking with the shield so postpone restoring
            postponeRestoring(p);
        else {
            p.getInventory().setItemInOffHand(storedOffhandItems.get(id));
            storedOffhandItems.remove(id);
        }
    }

    private void tryCancelTask(UUID id){
        RunnableSeries series = correspondingTasks.remove(id);
        if(series != null) series.cancelAll();
    }

    private void postponeRestoring(Player p){
        UUID id = p.getUniqueId();
        Optional.ofNullable(correspondingTasks.get(id))
                .ifPresent(RunnableSeries::cancelAll);

        correspondingTasks.remove(id);
        scheduleRestore(p);
    }

    private boolean isBlocking(UUID uuid){
        return storedOffhandItems.containsKey(uuid);
    }

    private boolean hasShield(Player p){
        return p.getInventory().getItemInOffHand().getType() == Material.SHIELD;
    }

    private boolean isHoldingSword(Material mat){
        return mat.toString().endsWith("_SWORD");
    }
}

File: src/main/kotlin/kernitus/plugin/OldCombatMechanics/module/ModuleSwordBlocking.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package kernitus.plugin.OldCombatMechanics.module

import kernitus.plugin.OldCombatMechanics.OCMMain
import kernitus.plugin.OldCombatMechanics.utilities.reflection.VersionCompatUtils
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockCanBuildEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.player.*
import org.bukkit.inventory.EquipmentSlot
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.PlayerInventory
import org.bukkit.scheduler.BukkitTask
import java.util.*

class ModuleSwordBlocking(plugin: OCMMain) : OCMModule(plugin, "sword-blocking") {
    // Not using WeakHashMaps here, for extra reliability
    private val storedItems: MutableMap<UUID, ItemStack> = HashMap()
    private val correspondingTasks: MutableMap<UUID, Collection<BukkitTask>> = HashMap()
    private var restoreDelay = 0

    // Only used <1.13, where BlockCanBuildEvent.getPlayer() is not available
    private var lastInteractedBlocks: MutableMap<Location, UUID>? = null

    companion object {
        private val SHIELD = ItemStack(Material.SHIELD)
    }

    init {
        if (!VersionCompatUtils.versionIsNewerOrEqualTo(1, 13, 0)) {
            lastInteractedBlocks = WeakHashMap()
        }
    }

    override fun reload() {
        restoreDelay = module().getInt("restoreDelay", 40)
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onBlockPlace(e: BlockCanBuildEvent) {
        if (e.isBuildable) return

        // If <1.13 get player who last interacted with block
        val player = lastInteractedBlocks?.remove(e.block.location)?.let(Bukkit::getPlayer) ?: e.player

        if (player == null || !isEnabled(player)) return

        doShieldBlock(player)
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onRightClick(e: PlayerInteractEvent) {
        val action = e.action
        val player = e.player

        if (!isEnabled(player)) return

        if (action != Action.RIGHT_CLICK_BLOCK && action != Action.RIGHT_CLICK_AIR) return
        // If they clicked on an interactive block, the 2nd event with the offhand won't fire
        // This is also the case if the main hand item was used, e.g. a bow
        // TODO right-clicking on a mob also only fires one hand
        if (action == Action.RIGHT_CLICK_BLOCK && e.hand == EquipmentSlot.HAND) return
        if (e.isBlockInHand) {
            e.clickedBlock?.location?.let { location ->
                lastInteractedBlocks?.set(location, player.uniqueId)
            }
            return  // Handle failed block place in separate listener
        }

        doShieldBlock(player)
    }

    private fun doShieldBlock(player: Player) {
        val inventory = player.inventory

        val mainHandItem = inventory.itemInMainHand
        val offHandItem = inventory.itemInOffHand

        if (!isHoldingSword(mainHandItem.type)) return

        if (module().getBoolean("use-permission") && !player.hasPermission("oldcombatmechanics.swordblock")) return

        val id = player.uniqueId

        if (!isPlayerBlocking(player)) {
            if (hasShield(inventory)) return
            debug("Storing $offHandItem", player)
            storedItems[id] = offHandItem

            inventory.setItemInOffHand(SHIELD)
            // Force an inventory update to avoid ghost items
            player.updateInventory()
        }
        scheduleRestore(player)
    }

    @EventHandler
    fun onHotBarChange(e: PlayerItemHeldEvent) = restore(e.player, true)

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onWorldChange(e: PlayerChangedWorldEvent) = restore(e.player, true)

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onPlayerLogout(e: PlayerQuitEvent) = restore(e.player, true)

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onPlayerDeath(e: PlayerDeathEvent) {
        val p = e.entity
        val id = p.uniqueId
        if (!areItemsStored(id)) return

        //TODO what if they legitimately had a shield?
        e.drops.replaceAll { item: ItemStack -> if (item.type == Material.SHIELD) storedItems.remove(id) else item }

        // Handle keepInventory = true
        restore(p, true)
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onPlayerSwapHandItems(e: PlayerSwapHandItemsEvent) {
        if (areItemsStored(e.player.uniqueId)) e.isCancelled = true
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onInventoryClick(e: InventoryClickEvent) {
        if (e.whoClicked is Player) {
            val player = e.whoClicked as Player

            if (areItemsStored(player.uniqueId)) {
                val cursor = e.cursor
                val current = e.currentItem
                if (cursor != null && cursor.type == Material.SHIELD || current != null && current.type == Material.SHIELD) {
                    e.isCancelled = true
                    restore(player, true)
                }
            }
        }
    }

    @EventHandler(priority = EventPriority.HIGHEST)
    fun onItemDrop(e: PlayerDropItemEvent) {
        val itemDrop = e.itemDrop
        val p = e.player

        if (areItemsStored(p.uniqueId) && itemDrop.itemStack.type == Material.SHIELD) {
            e.isCancelled = true
            restore(p)
        }
    }

    private fun restore(p: Player, force: Boolean = false) {
        val id = p.uniqueId

        tryCancelTask(id)

        if (!areItemsStored(id)) return

        // If they are still blocking with the shield, postpone restoring
        if (!force && isPlayerBlocking(p)) scheduleRestore(p)
        else p.inventory.setItemInOffHand(storedItems.remove(id))
    }

    private fun tryCancelTask(id: UUID) {
        correspondingTasks.remove(id)?.forEach { it.cancel() }
    }

    private fun scheduleRestore(p: Player) {
        val id = p.uniqueId
        tryCancelTask(id)

        val removeItem = Bukkit.getScheduler().runTaskLater(plugin, Runnable { restore(p) }, restoreDelay.toLong())

        val checkBlocking = Bukkit.getScheduler().runTaskTimer(plugin, Runnable {
            if (!isPlayerBlocking(p)) restore(p)
        }, 10L, 2L)

        val tasks: MutableList<BukkitTask> = ArrayList(2)
        tasks.add(removeItem)
        tasks.add(checkBlocking)
        correspondingTasks[p.uniqueId] = tasks
    }

    private fun areItemsStored(uuid: UUID) = storedItems.containsKey(uuid)

    /**
     * Checks whether player is blocking or they have just begun to and shield is not fully up yet.
     */
    private fun isPlayerBlocking(player: Player): Boolean {
        return player.isBlocking || (VersionCompatUtils.versionIsNewerOrEqualTo(
            1, 11, 0
        ) && player.isHandRaised && hasShield(player.inventory))
    }

    private fun hasShield(inventory: PlayerInventory) = inventory.itemInOffHand.type == Material.SHIELD

    private fun isHoldingSword(mat: Material) = mat.toString().endsWith("_SWORD")
}
